### ISSUE TITLE
Slow accuracy metric

### ISSUE THREAD

**Issue by Diuven** [PyTorchLightning/pytorch-lightning#2722]
## 🐛 Bug

Accuracy class from metrics.classification is too slow, compared to pytorch vanilla implementation.

### To Reproduce

Try the following demo code.

#### Code sample

```python3
from time import time
import torch
from pytorch_lightning.metrics.classification import Accuracy


def test_acc(pred, targ, get_acc, steps=1000, rep=10):
    tsum, mean_res = 0, []
    for i in range(rep):
        res = []
        ts = time()
        for j in range(steps):
            res.append(get_acc(pred, targ))
        te = time()
        tsum += te - ts
        mean_acc = torch.stack(res).mean().item()
        mean_res.append(mean_acc)
    
    return (tsum / rep), mean_res


def demo(batch_size=32, num_classes=10, device='cuda', steps=100, rep=5):
    print("Testing  bs: %d,  num_cls: %d,  device: %s,  steps: %d,  rep: %d" % (batch_size, num_classes, device, steps, rep))

    pred = torch.randint(0, num_classes, (batch_size, )).to(device=device)
    targ = torch.randint(0, num_classes, (batch_size, )).to(device=device)

    pl_acc = Accuracy(num_classes=num_classes)
    def pt_acc(pred, targ):
        return pred.eq(targ).to(dtype=torch.float).mean()

    pl_time, pl_res = test_acc(pred, targ, pl_acc, steps, rep)
    pt_time, pt_res = test_acc(pred, targ, pt_acc, steps, rep)

    if pl_res != pt_res:
        print("Results mismatch!")
    print("Lightning: %2.6fs" % pl_time)
    print("Vanilla  : %2.6fs" % pt_time)
    print("")

if __name__ == "__main__":
    demo(32, 10)
    demo(512, 10)
    demo(32, 100)
    demo(512, 100)
    demo(32, 1000)
    demo(512, 1000)

    demo(32, 10, 'cpu')
    demo(512, 10, 'cpu')
    demo(32, 100, 'cpu')
    demo(512, 100, 'cpu')
    demo(32, 1000, 'cpu')
    demo(512, 1000, 'cpu')
```

On my machine, the results are like the following.

```
Testing  bs: 32,  num_cls: 10,  device: cuda,  steps: 100,  rep: 5
Lightning: 0.378017s
Vanilla  : 0.002766s

Testing  bs: 512,  num_cls: 10,  device: cuda,  steps: 100,  rep: 5
Lightning: 0.372391s
Vanilla  : 0.002776s

Testing  bs: 32,  num_cls: 100,  device: cuda,  steps: 100,  rep: 5
Lightning: 3.346141s
Vanilla  : 0.002812s

Testing  bs: 512,  num_cls: 100,  device: cuda,  steps: 100,  rep: 5
Lightning: 3.335049s
Vanilla  : 0.002660s

Testing  bs: 32,  num_cls: 1000,  device: cuda,  steps: 100,  rep: 5
Lightning: 32.751692s
Vanilla  : 0.002666s

Testing  bs: 512,  num_cls: 1000,  device: cuda,  steps: 100,  rep: 5
Lightning: 32.967594s
Vanilla  : 0.002762s

Testing  bs: 32,  num_cls: 10,  device: cpu,  steps: 100,  rep: 5
Lightning: 0.128131s
Vanilla  : 0.001317s

Testing  bs: 512,  num_cls: 10,  device: cpu,  steps: 100,  rep: 5
Lightning: 0.128481s
Vanilla  : 0.001403s

Testing  bs: 32,  num_cls: 100,  device: cpu,  steps: 100,  rep: 5
Lightning: 1.117403s
Vanilla  : 0.001275s

Testing  bs: 512,  num_cls: 100,  device: cpu,  steps: 100,  rep: 5
Lightning: 1.166188s
Vanilla  : 0.001330s

Testing  bs: 32,  num_cls: 1000,  device: cpu,  steps: 100,  rep: 5
Lightning: 10.927701s
Vanilla  : 0.001229s

Testing  bs: 512,  num_cls: 1000,  device: cpu,  steps: 100,  rep: 5
Lightning: 11.560013s
Vanilla  : 0.001310s
```

100x ~ 10000x more time is needed for pl's accuracy, depending on the num_classes argument.

### Expected behavior

Pytorch-lightning's accuracy function should take similar time compared to the pytorch vanilla implementation, and it should not depend on num_classes too much.

### Environment

```
* CUDA:
	- GPU:
		- GeForce GTX 1050 Ti
	- available:         True
	- version:           10.2
* Packages:
	- numpy:             1.19.0
	- pyTorch_debug:     False
	- pyTorch_version:   1.5.1
	- pytorch-lightning: 0.8.1
	- tensorboard:       2.2.2
	- tqdm:              4.47.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.6.9
	- version:           #46~18.04.1-Ubuntu SMP Fri Jul 10 07:21:24 UTC 2020
```

I tested the same code from two other machines (each with 2080 ti and with V100), and they give similar results.

### Additional context

I'm not sure this issue falls into the bug report or the feature request category. Please relocate this as you like. :)

**Comment by Diuven** [comment]
Eh, now this issue looks a lot like a enhancement issue. Sorry :sweat_smile: 

I saw the implementation of the accuracy function, and I noticed a few metric functions depend on the [stat_scores_multiple_classes](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/metrics/functional/classification.py#L137) function.

I think this function is way too slow, as it doesn't run parallel on num_classes. It would be great if we can reimplement this function only with vanilla pytorch operations, but I can't find a way for that.

Considering the huge complexity gap, I think implementing on ad-hoc basis for each metric would be better, though it might be a bit ugly.

**Comment by justusschock** [comment]
Hi @Diuven , I was able to reproduce your results. 

I agree, that this is a huge slowdown and we should fix it. Unfortunately, I am not quite sure, why this happens.

I have profiled the code and by far the most time is taken to compute the stat scores.  Although I don't know why this takes that long in PyTorch, I am afraid, that we need a more general version. Do you have a proposal on how to parallelise the computation of stat scores among classes?

To be honest I don't like to duplicate stuff there. But I agree, that we should speed the current implementation up :)

**Comment by Diuven** [comment]
Hi @justusschock , thanks for your reply.

Yeah, I agree on not having duplicate codes in pl. It would be great if we simply replace `stat_scores_multiple_classes` with faster version.

Though I haven't profiled the code, I think the main reason is that there is no parallelism on `num_classes`. In current implementation of `stat_scores_multiple_classes`, we simply run for loops to compute states for each classes.
In the results of the demo code, the elapsed time also heavily depends on `num_classes`, so it supports the claim above.

So if we replace the for loop, `for c in range(num_classes):`,  in the function with the parallelized code, it would be much faster. Or at least it wouldn't depend on `num_classes`.


Making a CUDA library just for this function would work, but that would be our last option, considering the pl's philosophy and current state. (I think there's no CUDA code in this repo, so it would be an overkill to make a CUDA library only for this)

I'm not sure about whether this will work well or not, but I think parallelized [scatter](https://pytorch.org/docs/stable/tensors.html?highlight=scatter#torch.Tensor.scatter) function in pytorch might help.

**Comment by justusschock** [comment]
You're right that this would be great. Mind putting together a draft? I would for sure help you with that. 

Also it's worth mentioning, that PLs version will always be a bit slower than the pure pytorch version, since it does a bit more. I also checks all incoming and outgoing arguments for device and type and converts them if necessary :)

**Comment by Diuven** [comment]
Here's my draft:

```python3
def stat_scores_multiple_classes(pred: LongTensor, targ: LongTensor):
  match_true = (pred == targ).long()
  match_false = match_true.logical_not()
  # tps, fps, tns, fns, sups initialized as same
  pred_true = pred * match
  pred_false = pred - pred_true
  
  tps.scatter_add_(dim=0, pred, match_true)
  fps.scatter_add_(dim=0, pred, match_false)
  fns.scatter_add_(dim=0, targ, match_false)
  tns = pred.size(0) - tps - fps - fns
  sups.scatter_add_(dim=0, targ, torch.ones_like(match_true))

  return tps, fps, tns, fns, sups
```

Please check if I missed something big there. If it seems right, I'll work on it and make pull request.


> Also it's worth mentioning, that PLs version will always be a bit slower than the pure pytorch version, since it does a bit more. I also checks all incoming and outgoing arguments for device and type and converts them if necessary :)

Yeah, that's a really great part of PL, and why I like it. :smile_cat: I agree on compromising a bit of speediness to get huge convenience. All we need is just comparable performance to vanilla case, in my opinion.

**Comment by Diuven** [comment]
I checked the code below with the demo code at the top, and it was 10x faster than the previous code, but still slower than the vanilla code and still depends on the `num_classes`.

```python3
def stat_scores_multiple_classes(
        pred: torch.Tensor,
        target: torch.Tensor,
        num_classes: Optional[int] = None,
        argmax_dim: int = 1,
) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
    
    
    #num_classes = get_num_classes(pred=pred, target=target,
    #                              num_classes=num_classes)

    #if pred.ndim == target.ndim + 1:
    #    pred = to_categorical(pred, argmax_dim=argmax_dim)
    
    placeholder = torch.zeros((num_classes,), device=pred.device)

    match_true = (pred == target).float()
    match_false = 1 - match_true
    # tps, fps, tns, fns, sups initialized as same

    tps = placeholder.scatter_add(0, pred, match_true)
    fps = placeholder.scatter_add(0, pred, match_false)
    fns = placeholder.scatter_add(0, target, match_false)
    tns = pred.size(0) - (tps + fps + fns)
    sups = placeholder.scatter_add(0, target, torch.ones_like(match_true))

    return tps, fps, tns, fns, sups
```

I think if we don't give up the `(num_classes, )` sized tensor results, this is as far as we can get. If we make another version of `stat_scores_multiple_classes` returning only sums of the stats, it would be far better, in my opinion.

I'm not sure about whether we can bear with those two similar codes (sum version and tensor version of `stat_scores_multiple_classes` function). What's your thoughts?


Results:

```
Testing  bs: 32,  num_cls: 10,  device: cuda,  steps: 100,  rep: 5
Lightning: 0.053924s
Vanilla  : 0.002690s

Testing  bs: 512,  num_cls: 10,  device: cuda,  steps: 100,  rep: 5
Lightning: 0.052600s
Vanilla  : 0.002591s

Testing  bs: 32,  num_cls: 100,  device: cuda,  steps: 100,  rep: 5
Lightning: 0.234717s
Vanilla  : 0.002604s

Testing  bs: 512,  num_cls: 100,  device: cuda,  steps: 100,  rep: 5
Lightning: 0.237395s
Vanilla  : 0.002614s

Testing  bs: 32,  num_cls: 1000,  device: cuda,  steps: 100,  rep: 5
Lightning: 1.996236s
Vanilla  : 0.002550s

Testing  bs: 512,  num_cls: 1000,  device: cuda,  steps: 100,  rep: 5
Lightning: 2.008798s
Vanilla  : 0.002579s

Testing  bs: 32,  num_cls: 10,  device: cpu,  steps: 100,  rep: 5
Lightning: 0.019533s
Vanilla  : 0.001189s

Testing  bs: 512,  num_cls: 10,  device: cpu,  steps: 100,  rep: 5
Lightning: 0.020818s
Vanilla  : 0.001329s

Testing  bs: 32,  num_cls: 100,  device: cpu,  steps: 100,  rep: 5
Lightning: 0.080327s
Vanilla  : 0.001167s

Testing  bs: 512,  num_cls: 100,  device: cpu,  steps: 100,  rep: 5
Lightning: 0.081321s
Vanilla  : 0.001278s

Testing  bs: 32,  num_cls: 1000,  device: cpu,  steps: 100,  rep: 5
Lightning: 0.682839s
Vanilla  : 0.001193s

Testing  bs: 512,  num_cls: 1000,  device: cpu,  steps: 100,  rep: 5
Lightning: 0.704203s
Vanilla  : 0.001301s
```

**Comment by justusschock** [comment]
That's a great improvement :) 

As long as we only end up with these two alternatives and not one version per metric this should be fine, I think. But doesn't that also depend on the reduction type?

**Comment by Diuven** [comment]
Yeah you're right.

But as long as we use only `mean`, `sum` and `none` reduction methods, it the `stat_scores_multiple_classes` function won't be too long.

If we want to support additional reduction methods in the future, we probably can use `none` version of the function (i.e. function returning `(num_classes, )` sized tensors) and simply reduce after that. But if we have a lot faster version for certain reduction types, such as `sum` and `mean`, it would be very useful to have faster version of it.

**Comment by justusschock** [comment]
Great, do you want to just open a PR with your version and the even faster one?

**Comment by Diuven** [comment]
Yeah, sure. Small notes before that:

I changed and benchmarked it again in my forked repo: [changed function](https://github.com/Diuven/pytorch-lightning/blob/master/pytorch_lightning/metrics/functional/classification.py#L137)

In that file, note that I also slightly changed [accuracy function](https://github.com/Diuven/pytorch-lightning/blob/master/pytorch_lightning/metrics/functional/classification.py#L222) too. Please check if it has any wrong parts.

This gives around 1000x faster results than before, and it's only 10x slower than the vanilla implementation.

By the way, I curious about the difference of expected behaviour of `Accuracy` class when `reduction = 'elementwise_mean'` and  `reduction = 'sum'`. I think they are supposed to give same results. Am I right?

```
Testing  bs: 32,  num_cls: 10,  device: cuda,  steps: 100,  rep: 5
Lightning: 0.029696s
Vanilla  : 0.002814s

Testing  bs: 512,  num_cls: 10,  device: cuda,  steps: 100,  rep: 5
Lightning: 0.029949s
Vanilla  : 0.002888s

Testing  bs: 32,  num_cls: 100,  device: cuda,  steps: 100,  rep: 5
Lightning: 0.030695s
Vanilla  : 0.002683s

Testing  bs: 512,  num_cls: 100,  device: cuda,  steps: 100,  rep: 5
Lightning: 0.030434s
Vanilla  : 0.002817s

Testing  bs: 32,  num_cls: 1000,  device: cuda,  steps: 100,  rep: 5
Lightning: 0.030920s
Vanilla  : 0.002895s

Testing  bs: 512,  num_cls: 1000,  device: cuda,  steps: 100,  rep: 5
Lightning: 0.030054s
Vanilla  : 0.002917s

Testing  bs: 32,  num_cls: 10,  device: cpu,  steps: 100,  rep: 5
Lightning: 0.011744s
Vanilla  : 0.001375s

Testing  bs: 512,  num_cls: 10,  device: cpu,  steps: 100,  rep: 5
Lightning: 0.011817s
Vanilla  : 0.001461s

Testing  bs: 32,  num_cls: 100,  device: cpu,  steps: 100,  rep: 5
Lightning: 0.011544s
Vanilla  : 0.001259s

Testing  bs: 512,  num_cls: 100,  device: cpu,  steps: 100,  rep: 5
Lightning: 0.011770s
Vanilla  : 0.001412s

Testing  bs: 32,  num_cls: 1000,  device: cpu,  steps: 100,  rep: 5
Lightning: 0.012339s
Vanilla  : 0.001181s

Testing  bs: 512,  num_cls: 1000,  device: cpu,  steps: 100,  rep: 5
Lightning: 0.012118s
Vanilla  : 0.001297s
```

**Comment by justusschock** [comment]
@Djuven, great! Thanks a lot. From what I saw just now, your implementation and changes seem to be fine. can you open a PR to see if all our tests are still passing?

**Comment by Diuven** [comment]
Actually, I was running the existing tests in machine, but some tests are not passing.

I made a PR to share the results. Here's [the link](https://github.com/PyTorchLightning/pytorch-lightning/pull/2775) for the PR.

I think failed tests are the case when `num_classes` does not match with actual number of classes in the tensor. the `scatter_add` function throws an Error.

If we want to keep this functionality, rather than forcing the `num_classes` to be the actual number of classes, I'll make additional changes for that. Cutting of the out-of-bound values and some small fixes will do.

**Comment by awaelchli** [comment]
Cool! perhaps in the future the metrics package could include parity tests for speed, not only correctness.
